abm/projects/cooperative_signaling/cs_agent/cs_agent.py

```diff
@@ -54,93 +54,170 @@ def update(self, agents):
-            else:
-                # carry out movement accordingly
-                vel, theta = random_walk(desired_vel=self.max_exp_vel)
-                vel = (2 - self.velocity)
-                self.agent_type = "mars_miner"
-
-        # updating position accordingly
-        if not self.is_moved_with_cursor:  # we freeze agents when we move them
-            # updating agent's state variables according to calculated vel and
-            # theta
-            self.orientation += theta
-            # storing theta in short term memory for phototaxis
-            self.theta_prev = theta
-            self.prove_orientation()  # bounding orientation into 0 and 2pi
-            self.velocity += vel
-            # self.prove_velocity()  # possibly bounding velocity of agent
-
-            # new agent's position
-            new_pos = (
-                self.position[0] + self.velocity * np.cos(self.orientation),
-                self.position[1] - self.velocity * np.sin(self.orientation)
-            )
+    def calc_others_signaling_density_proj(self, agents):
+        # TODO: implement this correctly
+        signalling = [ag for ag in agents if ag.is_signaling]
+        return projection_field(
+            objects=signalling,
+            fov=self.FOV,
+            v_field_resolution=self.v_field_res,
+            position=tuple(self.position),
+            radius=self.radius,
+            orientation=self.orientation,
+            keep_distance_info=False,
+            non_expl_agents=None
+        )
 
-            # update the agent's position with constraints (reflection from the
-            # walls) or with the new position
-            self.position = list(self.reflect_from_walls(new_pos))
+    def update_state(self):
+        # update agent state based on the decision-making process
+        self.agent_state = agent_decision(
+            meter=self.meter,
+            max_signal_of_other_agents=self.others_signaling_density.max(
+                initial=0),
+            max_crowd_density=self.crowing_density.max(initial=0),
+            crowd_density_threshold=0.5)
+
+    def perform_action(self):
+        # update agent color
+        # TODO: remove this because it is executed in draw_update
+        # self.change_color()
+
+        # we freeze agents when we move them
+        if not self.is_moved_with_cursor:
+            # perform the agent's action according to the current state
+            if self.agent_state == "exploration":
+                self.exploration()
+            elif self.agent_state == "taxis":
+                self.taxis()
+            elif self.agent_state == "relocation":
+                self.relocation()
+            elif self.agent_state == "flocking":
+                self.flocking()
         else:
-            # self.agent_type = "signalling"
             print(self.meter)
 
+    def exploration(self):
+        vel, theta = random_walk(desired_vel=self.max_exp_vel)
+        # QUESTION: why is this 2?
```

Comment:
@mezdahun I am wondering why we do this: `vel = (2 - self.velocity)`

Reply:
So `vel` is the delta velocity in the given step. Currently if the velocity of the agent is not equal 2 it will be corrected. This should be parametrized to some desired velocity. In case we already get a vel from `random_walk` method, this correction is simply not necessary.

Reply:
Alright, I see, thank you for the explanation!